src/Stratis.Bitcoin.Features.Consensus.Tests/TimeSyncBehaviorTest.cs

```diff
@@ -28,173 +28,283 @@ public class TimeSyncBehaviorTest
+        }
+
+        /// <summary>
+        /// Checks that <see cref="TimeSyncBehaviorState.AddTimeData(IPAddress, TimeSpan, bool)"/>
+        /// forgets old samples if it has reached predefined limits.
+        /// </summary>
+        [Fact]
+        public void AddTimeData_WithLargeInboundSampleSet_ForgetsOldSamples()
+        {
+            var dateTimeProvider = new DateTimeProvider();
+            TimeSyncBehaviorState state = this.CreateTimeSyncBehaviorState(dateTimeProvider);
+
+            // add max inbound samples
+            for (int i = 0; i < TimeSyncBehaviorState.MaxInboundSamples; i++)
+            {
+                bool used = state.AddTimeData(new IPAddress(i), TimeSpan.FromSeconds(400), true);
                 Assert.True(used);
-                if (i >= outboundSamples - TimeSyncBehaviorState.MaxOutboundSamples) outSamples.Add(-i * 1000);
             }
 
-            var allSamples = new List<int>(inSamples);
-            for (int i = 0; i < TimeSyncBehaviorState.OutboundToInboundWeightRatio; i++)
-                allSamples.AddRange(outSamples);
+            // add enough outbound samples to get an adjustment - and across an even range
+            Assert.True(state.AddTimeData( new IPAddress(1), TimeSpan.FromSeconds(-500), false));
+            Assert.True(state.AddTimeData( new IPAddress(2), TimeSpan.FromSeconds(-250), false));
+            Assert.True(state.AddTimeData( new IPAddress(3), TimeSpan.FromSeconds(250), false));
+            Assert.True(state.AddTimeData( new IPAddress(4), TimeSpan.FromSeconds(500), false));
+
+            TimeSpan diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - 250*1000) < TimeEpsilonMs, $"should be 250ms because leaning towards the inbound value of 400ms, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
+
+            // add another batch of inbound with a greater offset than highest outbound
+            for (int i = TimeSyncBehaviorState.MaxInboundSamples; i < TimeSyncBehaviorState.MaxInboundSamples*2; i++)
+            {
+                Assert.True(state.AddTimeData(new IPAddress(i*2), TimeSpan.FromSeconds(-1000), true), $"index: {i}");
+            }
+
+            diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - (-250*1000)) < TimeEpsilonMs, $"should be -250ms because the new inbound replace the old and swing the median closer to the new -1000ms values, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
+        }
+         
+        /// <summary>
+        /// Sets up the system under test
+        /// </summary>
+        /// <param name="dateTimeProvider"></param>
+        /// <returns></returns>
+        private TimeSyncBehaviorState CreateTimeSyncBehaviorState(DateTimeProvider dateTimeProvider)
```

Comment:
Incomplete comments

Reply:
Can I remove params and returns? This is a helper method to create the subject under test, feels completely overkill to be documenting those parameters when they are commented to death elsewhere in the codebase.

Reply:
Actually we do not require comments in test at all, so the answer here would be yes you can delete that. In normal code however, omitting the comment would make the IntelliSense less usable